### PR TITLE
Fix retrieve command issues

### PR DESCRIPTION
**Pull request by smclay**
This is a quick fix but I think the `retrieve._handleCmdRetrieve()` function needs some further improvements. There are a few of issues:
1. The NGAS host ID may include the port value (e.g. `ngas01:7777`). There are few places in the code where we construct a URL (e.g. `host + ':' + port + '/' + command + ...`). This will often result in the URL `ngas01:7777:7777/RETRIEVE?...`. Somehow this still works most of the time in python 2 but the `urlparse` module in python 3 fails to parse the URL and throws an exception.
2. I think the `ngamsFileUtils.quickFileLocate()` function should really return a `NGAMS_HOST_CLUSTER` instead of `NGAMS_HOST_REMOTE` value. This makes more sense to me. Presently if the file is available in another node in the cluster it will end up being retrieved by the partner-sites routines. This will still work but it should really be handled as a redirect or as a proxy request depending on the config.
3. If the IP address is configured as `0.0.0.0` for all nodes in the cluster then the redirects will fail using this IP address
4. If the IP address is `0.0.0.0` then we have no alternative but to use the host ID. The problem with the host ID is that it does not contain the domain name. We should really construct a URL using the host ID, domain name and port address. The domain name is not always available as far as I can see.
Let me know what you think.

### CODE REVIEW

**Comment by smclay** [comment]
This is a much better PR. I have tested the different configuration options:
* enabling/disabling NgamsCfg.Server.ProxyMode
* enabling/disabling NgamsCfg.PartnerSites.ProxyMode
* setting NgamsCfg.Server.IpAddress to "0.0.0.0" or to a real IP address
This now works as expected.
You may have seen in the email conversation with Andreas that we considered the possibility of adding a configuration and database column for storing the FQDN for the NGAS node. This would make the code much simpler and more reliable when constructing a URL for another NGAS node. I will create a new branch for this eventually but at the moment I need to focus on some mirroring plugin issues.

File: src/ngamsServer/ngamsServer/commands/retrieve.py [hunk 1]
```diff
@@ -302,13 +302,13 @@ def _handleCmdRetrieve(srvObj,
...
 
-    # If still not located, try to contact associated NGAS sites to query
-    # if the file is available there.
-    # TODO:
-    if (not ipAddress):
-        pass
+    # If the hosts in the cluster have their IP address set to '0.0.0.0' this
+    # will break request. We need to construct the full host name instead.
+    host_address = ipAddress
+    if host_address == "0.0.0.0" or not host_address or host_address is None:
```
Comment (rtobar) [inline src/ngamsServer/ngamsServer/commands/retrieve.py:308]: `not host_address` already covers `host_address is None` so the latter can go

File: src/ngamsServer/ngamsServer/ngamsFileUtils.py [hunk 2]
```diff
@@ -97,6 +96,56 @@ def parse_host_id(host_id):
...
+
+    # Does the host address contain a domain?
+    # TODO: This is a crude way of checking by assuming that the domain always follows a '.'
+    if "." in host_address:
+        return host_address
+
+    if not domain or domain is None:
+        if location == NGAMS_HOST_LOCAL \
+            or location == NGAMS_HOST_CLUSTER \
+                or location == NGAMS_HOST_DOMAIN:
```
Comment (rtobar) [inline src/ngamsServer/ngamsServer/ngamsFileUtils.py:137]: Again, this could be simplified to:

```python
if not domain and location in (NGAMS_HOST_LOCAL, NGAMS_HOST_CLUSTER, NGAMS_HOST_DOMAIN)
```

File: src/ngamsServer/ngamsServer/ngamsFileUtils.py [hunk 3]
```diff
@@ -97,6 +96,56 @@ def parse_host_id(host_id):
...
+            domain = ngamsLib.getDomain()
+
+    if len(host_address) > 0 \
+        and (domain is not None and len(domain) > 0):
```
Comment (rtobar) [inline src/ngamsServer/ngamsServer/ngamsFileUtils.py:141]: I *think* this could also be simplified to:

```python
if host_address and domain:
```

File: src/ngamsServer/ngamsServer/ngamsFileUtils.py [hunk 4]
```diff
@@ -97,6 +96,56 @@ def parse_host_id(host_id):
...
+        return "{}.{}".format(host_address, domain)
+    elif len (host_address) > 0:
+        return "{}".format(host_address)
```
Comment (rtobar) [inline src/ngamsServer/ngamsServer/ngamsFileUtils.py:144]:
```python
elif host_address:
    return host_address
```

File: src/ngamsServer/ngamsServer/ngamsServer.py [hunk 5]
```diff
@@ -2392,7 +2391,7 @@ def _parse_and_run(args, prog, server_class):
         larg = arg.lower()
         if larg in _lower:
             if larg != arg:
-                print("WARNING: case-insenstive command-line option names are "
+                print("WARNING: case insensitive command-line option names are "
```
Comment (rtobar) [inline src/ngamsServer/ngamsServer/ngamsServer.py:2394]: The two changes in this file seem unrelated to this PR, could they be removed (hopefully through a rebase to leave no trace in the git history, otherwise I'll squash the commits when I merge)? I was expecting no changes in `ngmasServer.py` and these break that expectation.

**Review by rtobar — COMMENTED** [review]
Thanks @smclay and sorry for the delay; I see that you've been busy anyway pushing further changes so I hope my delayed response didn't block your work.

I think from the a general point of view changes look fine. As you said in the email thread, and this is something we've mentioned in the past too, networking, addressing and hostnames are a tricky business, so there're always be some unexpected corner case we're not dealing with correctly.

I've left a couple of comments mostly around code simplification, but otherwise I feel comfortable with the changes if they aren't breaking any tests, including the new one :smile:.

File: src/ngamsServer/ngamsServer/ngamsServer.py (hunk 5 above)
Comment (smclay) [reply src/ngamsServer/ngamsServer/ngamsServer.py:2394]: I small typo fix. It somehow slipped in with these changes. I am unsure how to undo it.

**Comment by smclay** [comment]
Thank you for the code review. I have simplified the code following your suggestions. Hopefully the merge can go through today. I would like to push out a new release for wider testing.

File: src/ngamsServer/ngamsServer/ngamsServer.py (hunk 5 above)
Comment (rtobar) [reply src/ngamsServer/ngamsServer/ngamsServer.py:2394]: OK, no problems, let's just leave it as is, it was a small detail anyway

**Comment by rtobar** [comment]
As soon as all tests pass I'll merge this in.
